// @pixi/ui/lib/MaskedFrame.js
'use strict';

var core = require('@pixi/core');
var graphics = require('@pixi/graphics');
var sprite = require('@pixi/sprite');
var view = require('./utils/helpers/view.js');

class MaskedFrame extends graphics.Graphics {
  constructor({ target, mask, borderWidth, borderColor }) {
    super();
    this.borderColor = borderColor;
    this.borderWidth = borderWidth;
    this.target = view.getView(target);
    this.targetMask = view.getView(mask);
    this.target.addChild(this.targetMask);
    this.target.mask = this.targetMask;
    if (borderWidth) {
      this.showBorder();
      this.target.x = borderWidth;
      this.target.y = borderWidth;
      const borderMask = typeof mask === "string" ? new sprite.Sprite(core.Texture.from(mask)) : mask.clone();
      borderMask.width += borderWidth * 2;
      borderMask.height += borderWidth * 2;
      this.mask = borderMask;
      this.addChild(borderMask);
    }
    this.addChild(this.target);
  }
  /** Shows a border. */
  showBorder() {
    this.beginFill(this.borderColor);
    const width = this.borderWidth * 2;
    this.drawRect(0, 0, this.target.width + width, this.target.height + width);
  }
  /** Hides a border. */
  hideBorder() {
    this.clear();
  }
}

exports.MaskedFrame = MaskedFrame;
